### ISSUE TITLE
Existing key not found shortly after the leader is restarted

### ISSUE THREAD

**Issue by motty-stratoscale** [hashicorp/consul#3649]
### `consul version` 
Server: (I have 3 servers, no clients)
```
Consul v0.8.4
Protocol 2 spoken by default, understands 2 to 3 (agent will automatically use protocol >2 when speaking to compatible agents)
```

### `consul info` 
Server: (I have 3 servers, no clients)
```
consul info
agent:
	check_monitors = 12
	check_ttls = 77
	checks = 91
	services = 92
build:
	prerelease = 
	revision = f436077
	version = 0.8.4
consul:
	bootstrap = false
	known_datacenters = 1
	leader = false
	leader_addr = 1.109.58.226:8300
	server = true
raft:
	applied_index = 828432
	commit_index = 828432
	fsm_pending = 0
	last_contact = 11.922805ms
	last_log_index = 828432
	last_log_term = 84
	last_snapshot_index = 823993
	last_snapshot_term = 84
	latest_configuration = [{Suffrage:Voter ID:1.109.0.5:8300 Address:1.109.0.5:8300} {Suffrage:Voter ID:1.109.2.141:8300 Address:1.109.2.141:8300} {Suffrage:Voter ID:1.109.58.226:8300 Address:1.109.58.226:8300}]
	latest_configuration_index = 2665
	num_peers = 2
	protocol_version = 2
	protocol_version_max = 3
	protocol_version_min = 0
	snapshot_version_max = 1
	snapshot_version_min = 0
	state = Follower
	term = 84
runtime:
	arch = amd64
	cpu_count = 12
	goroutines = 287
	max_procs = 12
	os = linux
	version = go1.8.3
serf_lan:
	coordinate_resets = 0
	encrypted = false
	event_queue = 0
	event_time = 76
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 107
	members = 3
	query_queue = 0
	query_time = 1
serf_wan:
	coordinate_resets = 0
	encrypted = false
	event_queue = 0
	event_time = 1
	failed = 0
	health_score = 0
	intent_queue = 0
	left = 0
	member_time = 6
	members = 3
	query_queue = 0
	query_time = 1
```

### Operating system and Environment details
uname -a  
`Linux stratonode2.node.strato 3.10.0-514.10.2.el7.strato0004.eac14abc4738.x86_64 #1 SMP Mon Jun 19 15:39:06 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux
`

### Description of the Issue (and unexpected/desired result)
Hi.
I recently saw a (rare?) issue, where a KV GET returns 404 on an existing key, shortly after the Consul process of the leader was restarted.
- "Shortly" is around 0.4-0.7 second.
- In the cases I saw, the leader that was restarted remained the leader.
- It can happen also with keys that were written several minutes before the disruption, during which the system has been stable (3 servers).

These points make me think that this inconsistency is a bug, and not a case of the "small window" mentioned in https://www.consul.io/api/index.html#default .

### Reproduction steps
(Doesn't always reproduce!)
Write some key, e.g. `consul kv put motmot2 mm`
Restart the consul process on the leader node, while running this loop on another node:
```
# for _ in {1..50}; do consul kv get motmot2; sleep 0.2; done
mm
mm
mm
mm
mm
mm
mm
mm
mm
Error querying Consul agent: Unexpected response code: 500
Error querying Consul agent: Unexpected response code: 500
Error! No key exists at: motmot2
Error! No key exists at: motmot2
Error! No key exists at: motmot2
Error! No key exists at: motmot2
mm
mm
mm
mm
mm
mm
mm
mm
mm
mm
mm
```
(You can use curl instead, and see that it returns 500 rightfully, but then 404 wrongfully..) 

### On the leader, when the restarted consul comes up
```
    2017/11/05 11:39:56 [WARN] raft: Heartbeat timeout from "" reached, starting election
    2017/11/05 11:39:56 [INFO] raft: Node at 1.109.0.5:8300 [Candidate] entering Candidate state in term 14
    2017/11/05 11:39:56 [INFO] raft: Election won. Tally: 2
    2017/11/05 11:39:56 [INFO] raft: Node at 1.109.0.5:8300 [Leader] entering Leader state
    2017/11/05 11:39:56 [INFO] raft: Added peer 1.109.2.141:8300, starting replication
    2017/11/05 11:39:56 [INFO] raft: Added peer 1.109.58.226:8300, starting replication
    2017/11/05 11:39:56 [INFO] consul: cluster leadership acquired
    2017/11/05 11:39:56 [INFO] consul: New leader elected: stratonode2
    2017/11/05 11:39:56 [INFO] raft: pipelining replication to peer {Voter 1.109.58.226:8300 1.109.58.226:8300}
    2017/11/05 11:39:56 [INFO] raft: pipelining replication to peer {Voter 1.109.2.141:8300 1.109.2.141:8300}
```
Most of the leader restarts look like this, whether the bug appears or not ;(
I expected to find errors with AppendEntries, old log, etc. , but the bug can appear without them as well.

### On the node that ran the queries
```
    2017/11/05 11:39:52 [ERR] http: Request GET /v1/kv/mancala/tablesVersions/distributionCurrentVersions/?recurse=None, error: rpc error: EOF from=127.0.0.1:65516
    2017/11/05 11:39:52 [ERR] http: Request GET /v1/kv/..., error: rpc error: EOF from=127.0.0.1:61356
    2017/11/05 11:39:52 [ERR] http: Request GET /v1/kv/..., error: rpc error: EOF from=127.0.0.1:61354
    ... (many reads that our system does)
    ...
    2017/11/05 11:39:52 [ERR] http: Request GET /v1/kv/..., error: rpc error: stream closed from=127.0.0.1:65516
    2017/11/05 11:39:52 [ERR] http: Request GET /v1/kv/motmot, error: rpc error: stream closed from=127.0.0.1:62514
    ...
    2017/11/05 11:39:53 [INFO] serf: EventMemberUpdate: stratonode2.stratocluster-10825318342979188457
    2017/11/05 11:39:56 [WARN] raft: Heartbeat timeout from "1.109.0.5:8300" reached, starting election
    2017/11/05 11:39:56 [INFO] raft: Node at 1.109.2.141:8300 [Candidate] entering Candidate state in term 14
    2017/11/05 11:39:56 [ERR] raft: Failed to make RequestVote RPC to {Voter 1.109.0.5:8300 1.109.0.5:8300}: EOF
    2017/11/05 11:39:56 [INFO] raft: Duplicate RequestVote for same term: 14
    2017/11/05 11:39:56 [INFO] raft: Node at 1.109.2.141:8300 [Follower] entering Follower state (Leader: "1.109.0.5:8300")
    2017/11/05 11:39:56 [INFO] consul: New leader elected: stratonode2
```

I'll produce more details in a few days, if you want.
Thanks in advance!

**Comment by slackpad** [comment]
Hi @motty-stratoscale this was fixed in Consul 0.8.5 by way of https://github.com/hashicorp/consul/issues/2644! It will attempt to retry for a while and then you'll get a 500 error `Not ready to serve consistent reads` if it takes too long.

**Comment by slackpad** [comment]
I'll close this out as a duplicate of #2644 but please let us know if you see this in a newer version of Consul.
